### PR TITLE
wasi: backfills constants and TODOs for remaining functions

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This also normalizes naming conventions

### CODE REVIEW

File: wasi/strings.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	"unicode/utf8"
+)
+
+// nullTerminatedStrings holds null-terminated strings. It ensures that
+// its length and total buffer size don't exceed the max of uint32.
+// nullTerminatedStrings are convenience struct for args_get and environ_get. (environ_get is not implemented yet)
+//
+// A Null-terminated string is a byte string with a NULL suffix ("\x00").
+// See https://en.wikipedia.org/wiki/Null-terminated_string
+type nullTerminatedStrings struct {
```
Comment (codefromthecrypt) [inline wasi/strings.go:14]: this and tests are just moved to declutter wasi.go which will become massive

File: wasi/types.go [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+	}
+	return fmt.Sprintf("errno(%d)", uint32(err))
+}
+
+// Note: Below prefers POSIX symbol names over WASI ones, even if the docs are from WASI.
+// See https://linux.die.net/man/3/errno
+// See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#variants-1
+const (
+	// ESUCCESS No error occurred. System call completed successfully.
+	ESUCCESS Errno = iota
```
Comment (codefromthecrypt) [inline wasi/types.go:25]: oops forgot to rename these. will do that

File: wasi/wasi.go [hunk 3]
```diff
@@ -3,72 +3,42 @@ package wasi
...
+// parameter is a memory offset to write the result to. As memory offsets are uint32, each parameter representing a
+// result is uint32.
 //
 // Note: In WebAssembly 1.0 (MVP), there may be up to one Memory per store, which means the wasm.HostFunctionCallContext
 // Memory is always the wasm.Store Memories index zero: `store.Memories[0].Buffer`
 //
 // See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md
 // See https://www.w3.org/TR/wasm-core-1/#memory-instances%E2%91%A0.
 type API interface {
-	// ArgsSizesGet is a WASI function that reads command-line argument data (Args) sizes.
```
Comment (codefromthecrypt) [inline wasi/wasi.go:27]: reordered to match docs

File: wasi/wasi.go [hunk 4]
```diff
@@ -101,65 +144,99 @@ type api struct {
 
 func (w *api) register(store *wasm.Store) (err error) {
 	for _, wasiName := range []string{
-		wasiUnstableName,
+		wasiUnstableName, // TODO: check if there are any signature incompatibility between stable and preview 1
 		wasiSnapshotPreview1Name,
 	} {
-		err = store.AddHostFunction(wasiName, "proc_exit", reflect.ValueOf(proc_exit))
```
Comment (codefromthecrypt) [inline wasi/wasi.go:107]: reordered to match docs

File: wasi/wasi.go [hunk 5]
```diff
@@ -101,65 +144,99 @@ type api struct {
 
 func (w *api) register(store *wasm.Store) (err error) {
```
Comment (mathetake) [inline wasi/wasi.go:145]: while here, let's make the `err = storere.AddHostFunction(...); err != nil {...}` a loop as well

**Review by mathetake — APPROVED** [review]
I will defer to @nullpo-head for the rest

File: wasi/wasi.go (hunk 5 above)
Comment (nullpo-head) [inline wasi/wasi.go:145]: if you don't mind, how about renaming these old `(w *api)` variable to `a` as well? They are remnant of `(w *wasiEnvironment)`

Comment (codefromthecrypt) [reply wasi/wasi.go:145]: sure

File: wasi/types.go [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+	ETIMEDOUT:       "ETIMEDOUT",
+	ETXTBSY:         "ETXTBSY",
+	EXDEV:           "EXDEV",
+	ENOTCAPABLE:     "ENOTCAPABLE",
+}
+
+const (
+	// FunctionArgsGet reads command-line argument data.
+	//
+	// In WebAssembly 1.0 (MVP) Text format, this signature is:
```
Comment (nullpo-head) [inline wasi/types.go:263]: Just curious. Does that help to have this signature comment in the doc of API interface in `wasi.go` side too?

**Review by nullpo-head — APPROVED** [review]
looks awesome. thanks for the work!

File: wasi/wasi.go (hunk 5 above)
Comment (codefromthecrypt) [reply wasi/wasi.go:145]: sure thing

File: wasi/types.go (hunk 6 above)
Comment (codefromthecrypt) [reply wasi/types.go:263]: I'll change the link so that we don't have to duplicate it, but it is more obvious where to go

**Comment by codefromthecrypt** [comment]
Thanks for the help here and especially behind the scenes @takaya!

**Pull request merged**